editor/editor_notes.cpp

```diff
@@ -0,0 +1,190 @@
+#include "editor/editor_notes.hpp"
+
+#include "platform/platform.hpp"
+
+#include "coding/internal/file_data.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "base/string_utils.hpp"
+#include "base/assert.hpp"
+#include "base/logging.hpp"
+#include "base/timer.hpp"
+
+#include "std/future.hpp"
+
+#include "3party/pugixml/src/pugixml.hpp"
+
+namespace
+{
+bool LoadFromXml(pugi::xml_document const & xml,
+                 vector<editor::Note> & notes,
+                 uint32_t & uploadedNotesCount)
+{
+  uint64_t notesCount;
+  auto const root = xml.child("notes");
+  if (!strings::to_uint64(root.attribute("count").value(), notesCount))
+    return false;
+  uploadedNotesCount = static_cast<uint32_t>(notesCount);
+  for (auto const xNode : root.select_nodes("note"))
+  {
+    m2::PointD point;
+
+    auto const node = xNode.node();
+    auto const point_x = node.attribute("x");
+    if (!point_x || !strings::to_double(point_x.value(), point.x))
+      return false;
+
+    auto const point_y = node.attribute("y");
+    if (!point_y || !strings::to_double(point_y.value(), point.y))
+      return false;
+
+    auto const text = node.attribute("text");
+    if (!text)
+      return false;
+
+    notes.emplace_back(point, text.value());
+  }
+  return true;
+}
+
+void SaveToXml(vector<editor::Note> const & notes,
+               pugi::xml_document & xml,
+               uint32_t const UploadedNotesCount)
+{
+  auto root = xml.append_child("notes");
+  root.append_attribute("count") = UploadedNotesCount;
+  for (auto const & note : notes)
+  {
+    auto node = root.append_child("note");
+    node.append_attribute("x") = DebugPrint(note.m_point.x).data();
+    node.append_attribute("y") = DebugPrint(note.m_point.y).data();
+    node.append_attribute("text") = note.m_note.data();
+  }
+}
+
+vector<editor::Note> MoveNoteVectorAtomicly(vector<editor::Note> && notes, mutex & mu)
```

Comment:
Please, inline this code to the place where it's used. I generally don't like the idea to pass mutexes to functions.